cli/dcos-spark/security.go

```diff
@@ -0,0 +1,236 @@
+const SECRET_REFERENCE_TEMPLATE = "spark.mesos.%s.secret.names"
+const SECRET_FILENAME_TEMPLATE = "spark.mesos.%s.secret.filenames"
+const SECRET_ENVKEY_TEMPLATE = "spark.mesos.%s.secret.envkeys"
+const sparkKdcHostnameKey = "SPARK_SECURITY_KERBEROS_KDC_HOSTNAME"
+const sparkKdcPortKey = "SPARK_SECURITY_KERBEROS_KDC_PORT"
+const sparkKerberosRealmKey = "SPARK_SECURITY_KERBEROS_REALM"
+const sparkKrb5ConfBase64BlobKey = "SPARK_MESOS_KRB5_CONF_BASE64"
+
+
+func prepareBase64Secret(secretPath string) string {
+	ss := strings.Split(secretPath, "/")
+	// The secret file without any slashes
+	fn := ss[len(ss) - 1]
+	// secrets with __dcos_base64__ will be decoded by Mesos, but remove the prefix here
+	if strings.HasPrefix(fn, "__dcos_base64__") {
+		return strings.TrimPrefix(fn, "__dcos_base64__")
+	}
+	return fn
+}
+
+func addArgsForFileBasedSecret(args *sparkArgs, secretPath, property string) {
+	taskTypes := []string{"driver", "executor"}
+	for _, taskType := range taskTypes {
+		secretRefProp := fmt.Sprintf(SECRET_REFERENCE_TEMPLATE, taskType)
+		secretFileProp := fmt.Sprintf(SECRET_FILENAME_TEMPLATE, taskType)
+		appendToProperty(secretRefProp, secretPath, args)
+		appendToProperty(secretFileProp, prepareBase64Secret(secretPath), args)
+	}
+	args.properties[property] = prepareBase64Secret(secretPath)
+}
+
+func setupTLSArgs(args *sparkArgs) {
+	args.properties["spark.mesos.containerizer"] = "mesos"
+	args.properties["spark.ssl.enabled"] = "true"
+
+	// Keystore and truststore
+	const keyStoreFileName = "server.jks"
+	const trustStoreFileName = "trust.jks"
+	args.properties["spark.ssl.keyStore"] = keyStoreFileName
+	if args.truststoreSecretPath != "" {
+		args.properties["spark.ssl.trustStore"] = trustStoreFileName
+	}
+
+	// Secret paths, filenames, and place holder envvars
+	paths := []string{args.keystoreSecretPath}
+	filenames := []string{keyStoreFileName}
+	envkeys := []string{"DCOS_SPARK_KEYSTORE"}
+	if args.truststoreSecretPath != "" {
+		paths = append(paths, args.truststoreSecretPath)
+		filenames = append(filenames, trustStoreFileName)
+		envkeys = append(envkeys, "DCOS_SPARK_TRUSTSTORE")
+	}
+	joinedPaths := strings.Join(paths, ",")
+	joinedFilenames := strings.Join(filenames, ",")
+	joinedEnvkeys := strings.Join(envkeys, ",")
+
+	taskTypes := []string{"driver", "executor"}
+	for _, taskType := range taskTypes {
+		appendToProperty(fmt.Sprintf(SECRET_REFERENCE_TEMPLATE, taskType), joinedPaths, args)
+		appendToProperty(fmt.Sprintf(SECRET_FILENAME_TEMPLATE, taskType), joinedFilenames, args)
+		appendToPropertyIfSet(fmt.Sprintf(SECRET_ENVKEY_TEMPLATE, taskType), joinedEnvkeys, args)
+	}
+
+	// Passwords
+	args.properties["spark.ssl.keyStorePassword"] = args.keystorePassword
+	args.properties["spark.ssl.keyPassword"] = args.privateKeyPassword
+
+	if args.truststoreSecretPath != "" {
+		args.properties["spark.ssl.trustStorePassword"] = args.truststorePassword
+	}
+
+	// Protocol
+	if _, ok := args.properties["spark.ssl.protocol"]; !ok {
```

Comment:
If we don't allow other settings for `spark.ssl.protocol`, then we should simply set it here instead of only setting it when empty. If other settings are allows, one should probably log overwrites or invalid settings here.

Reply:
We do. In Spark we don't (currently) automate the generation of the TLS artifacts. Therefore we allow the user to specify whichever they would like, but `TLS` as the default.